### PR TITLE
COE-753 Merge measurements and event_based_simulators

### PR DESCRIPTION
**Pull request by namnhatpham**

### CODE REVIEW

File: event-based-simulators/main/cumulocityAPI.py [hunk 1]
```diff
@@ -7,19 +7,28 @@
 
 MOCK_REQUESTS = os.environ.get('MOCK_C8Y_REQUESTS') or 'false'
 
-user_and_pass_bytes = base64.b64encode((C8Y_TENANT + "/" + C8Y_USER + ':' + C8Y_PASSWORD).encode('ascii')) # bytes
-user_and_pass = user_and_pass_bytes.decode('ascii') # decode to str 
+user_and_pass_bytes = base64.b64encode((C8Y_TENANT + "/" + C8Y_USER + ':' + C8Y_PASSWORD).encode('ascii'))  # bytes
+user_and_pass = user_and_pass_bytes.decode('ascii')  # decode to str
```
Comment (mvoigt-sag) [inline event-based-simulators/main/cumulocityAPI.py:11]: please revert unnecessary changes, so the PR doesnt get cluttered with irrelevant changes...

File: event-based-simulators/main/cumulocityAPI.py [hunk 2]
```diff
@@ -7,19 +7,28 @@
...
     'Content-Type': 'application/json',
     'Accept': 'application/json',
     'Authorization': 'Basic ' + user_and_pass
 }
-
+MEASUREMENTS_COLLECTION_HEADERS = {
+    'Content-Type': 'application/vnd.com.nsn.cumulocity.measurementcollection+json',
+    'Accept': 'application/json',
+    'Authorization': 'Basic ' + user_and_pass
+}
```
Comment (mvoigt-sag) [inline event-based-simulators/main/cumulocityAPI.py:22]: is MEASUREMENTS_COLLECTION_HEADERS used anywhere?

File: event-based-simulators/main/cumulocityAPI.py [hunk 3]
```diff
@@ -48,14 +69,16 @@ def get_or_create_device(self, sim_id, label):
         if self.mocking:
             log.info(f"mock: get or create device with external id {sim_id}")
             return sim_id
-        
+
```
Comment (mvoigt-sag) [inline event-based-simulators/main/cumulocityAPI.py:72]: see first comment rgd irrelevant changes; also applies to a few more changes below

File: event-based-simulators/main/cumulocityAPI.py [hunk 4]
```diff
@@ -7,19 +7,28 @@
...
     'Accept': 'application/json',
     'Authorization': 'Basic ' + user_and_pass
 }
-
+MEASUREMENTS_COLLECTION_HEADERS = {
+    'Content-Type': 'application/vnd.com.nsn.cumulocity.measurementcollection+json',
+    'Accept': 'application/json',
+    'Authorization': 'Basic ' + user_and_pass
+}
+MEASUREMENTS_HEADERS = {
```
Comment (mvoigt-sag) [inline event-based-simulators/main/cumulocityAPI.py:23]: maybe better name is "MEASUREMENT_HEADERS"?

File: event-based-simulators/main/event_based_simulators.py [hunk 5]
```diff
@@ -1,22 +1,26 @@
...
 oeeAPI = OeeAPI()
 
-#Get Tenant Options and configure Simulator
+# Get Tenant Options and configure Simulator
 microservice_options = cumulocityAPI.get_tenant_option_by_category("event-based-simulators")
 PROFILE_CREATE_MODE = ProfileCreateMode[microservice_options.get("CREATE_PROFILES", "CREATE_IF_NOT_EXISTS")]
 CREATE_PROFILES_ARGUMENTS = microservice_options.get("CREATE_PROFILES_ARGUMENTS", "")
 CREATE_ASSET_HIERACHY = microservice_options.get("CREATE_ASSET_HIERACHY", "False")
-LOG_LEVEL = microservice_options.get("LOG_LEVEL", "INFO")
+LOG_LEVEL = microservice_options.get("LOG_LEVEL", "DEBUG")
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:23]: I think for the release we want to keep the log level at INFO

File: event-based-simulators/main/event_based_simulators.py [hunk 6]
```diff
@@ -345,35 +277,42 @@ def __send_following_event(self, event_definition, timestamp = None, extra_param
...
+                f'{self.device_id} task({id(task)}) removed: {json.dumps(event_definition)}, tasks: {len(self.tasks)}')
+
     def tick(self):
         if not self.enabled: return
-        if not self.is_in_productionTime():
-            if not self.out_of_production_time_logged:
-                self.__log_not_in_shift()
-            self.out_of_production_time_logged = True
-            return
+        if self.current_work == "event":
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:301]: why is this only relevant for events?

File: event-based-simulators/main/event_based_simulators.py [hunk 7]
```diff
@@ -345,35 +277,42 @@ def __send_following_event(self, event_definition, timestamp = None, extra_param
...
+            if not self.is_in_productionTime():
+                if not self.out_of_production_time_logged:
+                    self.__log_not_in_shift()
+                self.out_of_production_time_logged = True
+                return
+            else:
+                self.out_of_production_time_logged = False
+            for task in self.tasks[0]:
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:309]: psl use a constant with a proper name for "0"?

File: event-based-simulators/main/event_based_simulators.py [hunk 8]
```diff
@@ -345,35 +277,42 @@ def __send_following_event(self, event_definition, timestamp = None, extra_param
...
+                task.tick()
         else:
-            self.out_of_production_time_logged = False
-        for task in self.tasks:
-            task.tick()
+            for task in self.tasks[1]:
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:312]: psl use a constant with a proper name for "1"

File: event-based-simulators/main/event_based_simulators.py [hunk 9]
```diff
@@ -386,22 +325,38 @@ def get_or_create_device_id(self):
...
-        task = PeriodicTask(min_interval_in_seconds, max_interval_in_seconds, event_callback)
-        
-        log.debug(f'create periodic task for {event_definition["type"]} ({min_frequency_per_hour}, {max_frequency_per_hour})')        
-        # event_callback()
         return task
-    
-    def __send_event(self, event_fragment, timestamp = None):
+
+    def lambda_functions(self, event_definition, measurement_definition):
+        if self.current_work == "event":
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:351]: use a constant for the strings "event" and "measurement" and use it everywhere where needed

File: event-based-simulators/main/event_based_simulators.py [hunk 10]
```diff
@@ -419,20 +374,91 @@ def __send_event(self, event_fragment, timestamp = None):
...
+                        start = datetime.strptime(timeslot.slotStart, Shiftplan.dateformat)
+                        end = datetime.strptime(timeslot.slotEnd, Shiftplan.dateformat)
                         if start < now and end > now:
                             return True
         return has_no_shiftplan
 
+    # Measurements functions #
+    def measurement_functions(self, measurement_definition, task):
+        MachineSimulator.generate_measurement(self=self, measurement_definition=measurement_definition)
+        MachineSimulator.send_create_measurements(self=self, measurement_definition=measurement_definition)
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:394]: the name "send_create_measurements" seems a bit odd; maybe "send_measurement" is better?

**Review by mvoigt-sag — CHANGES_REQUESTED** [review]
I added a few comments and here are two more thoughts:
- formatting:
-- it seems there are a lot of code changes due to some formatting; I'd prefer to have them in a different PR (at least next time) b/c this way it is not so easy to see the really relevant changes
-- please sync with Peer how to handle formatting in the future
- simulators.json
-- I thought we agreed to not change the existing simulators and add new ones for sending measurements?

File: event-based-simulators/main/event_based_simulators.py (hunk 6 above)
Comment (namnhatpham) [reply event-based-simulators/main/event_based_simulators.py:301]: inside this condition, there are only functions related to event such as is_in_productionTime(), out_of_production_time_logged(), etc. 
measurement generation doesn't need these lines to work

File: event-based-simulators/main/cumulocityAPI.py (hunk 2 above)
Comment (mvoigt-sag) [reply event-based-simulators/main/cumulocityAPI.py:22]: If it is not needed, it should be removed. We can add it, once it is required. Or do you think you will need it within this PR?

File: event-based-simulators/main/event_based_simulators.py (hunk 6 above)
Comment (mvoigt-sag) [reply event-based-simulators/main/event_based_simulators.py:307]: well, I think also measurements should only be sent during production time. @haraldmeyer @pejo-sag what do you think?

File: event-based-simulators/main/event_based_simulators.py [hunk 11]
```diff
@@ -1,23 +1,29 @@
...
-LOG_LEVEL = microservice_options.get("LOG_LEVEL", "INFO")
-DELETE_PROFILES = microservice_options.get("DELETE_PROFILES", "False")
+# Get Tenant Options and configure Simulator
+MICROSERVICE_OPTIONS = cumulocityAPI.get_tenant_option_by_category("event-based-simulators")
+PROFILE_CREATE_MODE = ProfileCreateMode[MICROSERVICE_OPTIONS.get("CREATE_PROFILES", "CREATE_IF_NOT_EXISTS")]
+CREATE_PROFILES_ARGUMENTS = MICROSERVICE_OPTIONS.get("CREATE_PROFILES_ARGUMENTS", "")
+CREATE_ASSET_HIERARCHY = MICROSERVICE_OPTIONS.get("CREATE_ASSET_HIERARCHY", "False")
+LOG_LEVEL = MICROSERVICE_OPTIONS.get("LOG_LEVEL", "INFO")
+DELETE_PROFILES = MICROSERVICE_OPTIONS.get("DELETE_PROFILES", "False")
+ACTIONS_LIST = ["event", "measurement"]
```
[the remark this answers is not in the document]
Comment (namnhatpham) [reply event-based-simulators/main/event_based_simulators.py:25]: Do you mean I should create a constant like EVENT = "event" and use it instead of directly "event"?

[the remark this answers is not in the document]
Comment (mvoigt-sag) [reply event-based-simulators/main/event_based_simulators.py:25]: yes. and use it whenever the same value is used again in the script.

File: event-based-simulators/main/event_based_simulators.py (hunk 6 above)
Comment (haraldmeyer) [reply event-based-simulators/main/event_based_simulators.py]: I agree with Michael. Measurements are just a different way in which the devices send the data, so the same logic should apply to measurements and events.

e.g. you can have either "pieces produced" events from which you extract a piece count or you can have a measurement that represents a piece count. In both cases the assumption is that these should only happen during production time.

File: event-based-simulators/main/event_based_simulators.py [hunk 12]
```diff
@@ -419,20 +384,88 @@ def __send_event(self, event_fragment, timestamp = None):
...
+        self.simulated_data.append({
+            'type': measurement_definition.get("type"),
+            'series': measurement_definition.get("series"),
+            'value': value,
+            'unit': measurement_definition.get("unit"),
+            'time': self.current_time,
+        })
+
+    def send_measurements(self, measurement_definition):
+        json_measurements_list = []
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:432]: why do we need to create an array if we only ever add a single element?

File: event-based-simulators/main/event_based_simulators.py [hunk 13]
```diff
@@ -419,20 +384,88 @@ def __send_event(self, event_fragment, timestamp = None):
...
         return has_no_shiftplan
 
+    # Measurements functions #
+    def measurement_functions(self, measurement_definition, task):
+        MachineSimulator.generate_measurement(self=self, measurement_definition=measurement_definition)
+        MachineSimulator.send_measurements(self=self, measurement_definition=measurement_definition)
+
+    def generate_measurement(self, measurement_definition):
+        log.info(f"Generating value of measurement {measurement_definition.get('series')} of device {self.model.get('id')}")
+        self.simulated_data = []
```
Comment (mvoigt-sag) [inline event-based-simulators/main/event_based_simulators.py:408]: why do we need to create an array if we only ever add a single element?

Comment (namnhatpham) [reply event-based-simulators/main/event_based_simulators.py:408]: I created this method when thinking about sending a collection of measurements but then I changed to send only a single one so it is still left like that. Since it doesn't affect the code until now so I forget about it

File: event-based-simulators/main/event_based_simulators.py (hunk 12 above)
Comment (namnhatpham) [reply event-based-simulators/main/event_based_simulators.py:432]: I created this method when thinking about sending a collection of measurements but then I changed to send only a single one so it is still left like that. Since it doesn't affect the code until now so I forget about it

File: event-based-simulators/main/event_based_simulators.py (hunk 13 above)
Comment (mvoigt-sag) [reply event-based-simulators/main/event_based_simulators.py:408]: > this simulated data array contains all the measurement definitions of a machine so it can have multiple values

are you sure? b/c I can see only one `.append()` invocation and a couple of accesses with `simulated_data[0]`

Comment (namnhatpham) [reply event-based-simulators/main/event_based_simulators.py:407]: After carefully checking, I realize what you found is correct. I will make modification about it also

**Review by mvoigt-sag — APPROVED** [review]
